### PR TITLE
[ANE-611] Unarchiving VSI scan

### PR DESCRIPTION
**Pull request by skilly-lily**
# Overview

This PR uses the internal `archive` library to recursively expand archives while performing a VSI scan.

## Acceptance criteria

- Archives are expanded and fingerprinted while running `vsi::scan::artifacts`.

## Testing plan

Added an automated integration test for this.  The data was stolen from the archive expansion tests.

## Risks

Some placeholder filter functionality was removed, as it now resides in the archive library, and will be implemented as an extension of the `archive::expand::walk` function.  This should be double-checked, as I was not sure this was entirely the correct move.

## References

ANE-611

## Checklist

- [ ] I added tests for this PR's change (or explained in the PR description why tests don't make sense).

### CODE REVIEW

File: archive/src/expand/walk.rs [hunk 1]
```diff
@@ -86,6 +86,15 @@ impl Entry {
         File::open(&self.concrete)
     }
 
+    /// Consume the `Entry` and return the canonical path relative to the expanding root.
+    ///
+    /// Note that this path is only usable for recording purposes, and the path is not
+    /// suitable for opening. Additionally, the true underlying path may be deleted once
+    /// the `Entry` is dropped, making this
+    pub fn into_path(self) -> PathBuf {
```
Comment (jssblck) [inline archive/src/expand/walk.rs]: Please finish this sentence 😄

File: vsi/tests/it/main.rs [hunk 2]
```diff
@@ -43,3 +51,81 @@ async fn run_succeeds() -> Result<()> {
...
+    async fn append_scan(&self, id: &Id, artifacts: Vec<Artifact>) -> Result<()> {
+        ensure!(id == &self.id, "id's did not match: {} <=> {}", id, self.id);
+        self.buffer.lock().await.extend(artifacts);
+        Ok(())
+    }
+}
+
+#[tokio::test]
+async fn archive_scan_produces_correct_prints() -> Result<()> {
+    let id = Id::new("Fake ID, I promise I'm 21.");
```
Comment (jssblck) [inline vsi/tests/it/main.rs:72]: 🤣

File: vsi/tests/it/main.rs [hunk 3]
```diff
@@ -43,3 +51,81 @@ async fn run_succeeds() -> Result<()> {
...
+
+    let result_map: HashMap<_, _> = results
+        .into_iter()
+        .map(|art| {
+            let (key, comb) = art.explode_string();
+            (key, comb.to_string())
+        })
+        .collect();
+
+    let pathsep = std::path::MAIN_SEPARATOR;
```
Comment (jssblck) [inline vsi/tests/it/main.rs:103]: 🤔 I think we actually should be replacing `\` with `/` in paths... maybe that's why the windows integration test is broken. Feel free to leave that though, we can figure that out in the ticket that we have for fixing windows scans.

File: vsi/tests/it/main.rs [hunk 4]
```diff
@@ -43,3 +51,81 @@ async fn run_succeeds() -> Result<()> {
...
+        simple_actual,
+        Some(simple_zip_expected),
+        "comparing simple.zip"
+    );
+
+    let a_actual = result_map
+        .get(&format!(
+            "simple.zip!_fossa.virtual_!{pathsep}simple{pathsep}a.txt"
+        ))
+        .map(|s| s.as_str());
```
Comment (jssblck) [inline vsi/tests/it/main.rs:118]: [optional] Consider making a convenience macro for these:
```
let a_actual = get_joined!(result_map, "simple.zip!_fossa.virtual_!", "simple", "a.txt");
```

**Review by jssblck — APPROVED** [review]

File: vsi/tests/it/main.rs (hunk 4 above)
Comment (skilly-lily) [reply vsi/tests/it/main.rs:118]: I tried this but I had some trouble figuring out an approach so I dropped it in favor of time.

Comment (jssblck) [reply vsi/tests/it/main.rs:118]: Fair enough!

**Pull request merged**